### PR TITLE
Use AMQP queue in tests-scan

### PR DESCRIPTION
**Pull request by Gundersanne**
todo after merging:
- [x] deploy webhook and amqp pod on bos
- [ ] redeploy task bots

### CODE REVIEW

File: .tasks [hunk 1]
```diff
@@ -13,8 +13,8 @@
 
 set -ex
 
-# Scan for all tests
-bots/tests-scan
+# Process exactly one item from the queue
+bots/tests-dequeue-one --amqp cockpitrabbitmq.cockpit.svc.cluster.local --queue tests
```
Comment (martinpitt) [inline .tasks]: bikeshed alert.. As the main purpose of this is to *run* a test (or other action), and dequeuing is the means, plus in the future this should do "issue-scan" like things, too -- how about naming this "bots/run-queue"?

I'm not sure if we should have separate queues for separate kinds of tasks (e. g "po-refresh", "image-refresh", "test"), or have a queue item JSON contain a type field, or just contain a completely opaque shell script. My gut feeling is that the latter is easiest for now, and have at most two queues (tasks and redhat-tasks). We can refine the queue structure later on, if that at some point becomes impractical. WDYT?

File: bots/tests-dequeue-one [hunk 2]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/python3
+# -*- coding: utf-8 -*-
```
Comment (martinpitt) [inline bots/tests-dequeue-one]: This isn't necessary for Py 3.

File: bots/tests-dequeue-one [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+
+MAX_PRIORITY = 12
+
+import argparse
+import pika
```
Comment (martinpitt) [inline bots/tests-dequeue-one]: Can you please move "pika" below into a separate block? It's maybe just a whim of mine, but I like to separate Python builtin modules from third-party extensions.

File: bots/tests-dequeue-one [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import subprocess
+import sys
+
+def main():
+    parser = argparse.ArgumentParser(description='Bot: read test commands from the queue and execute them')
+    parser.add_argument('--amqp', action="store_true", default='localhost',
+                        help='The url of the amqp server to consume from.')
+    parser.add_argument('--queue', action="store_true", default='tests',
+                        help='The name of the queue to consume from.')
```
Comment (martinpitt) [inline bots/tests-dequeue-one]: `store_true` is wrong for these, as they get a string argument, these are not boolean. `store` is the default, so just leave out `action`.

File: bots/tests-dequeue-one [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    channel.basic_qos(0, 1, True)
+    channel.basic_consume(run_test, queue=opts.queue)
+    channel.start_consuming()
+
+def run_test(ch, method, properties, body):
+    command = body.decode('utf-8')
+
+    try:
+        # exists zero on test completion (so not just on positive test result)
+        subprocess.check_call(command.split(' '))
```
Comment (martinpitt) [inline bots/tests-dequeue-one]: I don't think this works. First of all, tests-scan produces two commands, not one:

    
PRIORITY=0006 bots/make-checkout --verbose --base=master --repo=weldr/welder-web --bots-ref=pull/10716/head '' && TEST_PROJECT=weldr/welder-web TEST_NAME=pull-10716-20181130-170707 TEST_REVISION=4cd21429e3cb4fe7db97e518858f6f69106b0254 bots/tests-invoke --rebase=master --remote=test cockpit/fedora-29/chrome ''

And second, if you have quoted arguments that contain spaces, this simple-minded split would break them apart. I think it's better to just use `shell=True`.

File: bots/tests-dequeue-one [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        ch.basic_ack(delivery_tag=method.delivery_tag)
+    except:
```
Comment (martinpitt) [inline bots/tests-dequeue-one]: No bare except, please. Either intercept `subprocess.CalledProcessError`, or don't catch exceptions at all. I don't think you need to explicitly nack anyway -- if the process dies, AMQP will see the TCP connection reset and automatically requeue it.

File: bots/tests-invoke [hunk 7]
```diff
@@ -249,7 +249,7 @@ class PullTask(object):
                 self.github_status_data["state"] = "success"
         if isinstance(ret, str):
             message = ret
-            mark_failed()
+            ret = 0
```
Comment (martinpitt) [inline bots/tests-invoke]: The `ret = 0` seems correct, but why drop the `mark_failed()`? That sounds like it woudl not update the GitHub status accordingly, and thus the status would be "pending" forever?

File: bots/tests-scan [hunk 8]
```diff
@@ -118,6 +121,11 @@ def main():
                         help='Test contexts to use.')
     parser.add_argument('-p', '--pull-numbers', action="append", default=[ ],
                         help='Pull requests to scan for tasks')
+    parser.add_argument('--amqp', action="store_true", default='localhost',
+                        help='The url of the amqp server to publish to.')
+    parser.add_argument('--queue', action="store_true", default='tests',
+                        help='The name of the queue to consume from.')
```
Comment (martinpitt) [inline bots/tests-scan]: Same thing again, these are string value, not boolean options. Also, this is not "consume from", but "publish to".

File: bots/tests-scan [hunk 9]
```diff
@@ -390,9 +421,10 @@ def scan_for_pull_tasks(api, policy, opts, repo):
     if opts.human_readable:
         func = lambda x: tests_human(*x)
         results.sort(reverse=True, key=lambda x: str(x))
-    else:
```
Comment (martinpitt) [inline bots/tests-scan]: The `else:` was accidentally (?) removed. "human view" is for, well, interactive human test descriptions. These should *not* get queued. You can check this locally with `bots/tests-scan -vd`, it should show running and pending tests.

File: bots/tests-scan [hunk 10]
```diff
@@ -118,6 +121,11 @@ def main():
                         help='Test contexts to use.')
     parser.add_argument('-p', '--pull-numbers', action="append", default=[ ],
                         help='Pull requests to scan for tasks')
+    parser.add_argument('--amqp', action="store_true", default='localhost',
```
Comment (martinpitt) [inline bots/tests-scan]: "URL" and "AMQP" should be capitalized.

File: bots/tests-scan [hunk 11]
```diff
@@ -163,8 +183,22 @@ def tests_human(priority, name, revision, ref, context, base, repo, bots_ref):
         bots_ref=bots_ref and (" [bots@%s]" % bots_ref) or "",
     )
 
+def queue_test(priority, name, revision, ref, context, base, repo, bots_ref, queue, options):
+    command = prepare_invoke(priority, name, revision, ref, context, base, repo, bots_ref, options)
+    if not command:
+        return
+
+    command = "echo 'hai'"
```
Comment (martinpitt) [inline bots/tests-scan]: `echo "I am a test!"` :wink:

File: bots/tests-scan [hunk 12]
```diff
@@ -163,8 +183,22 @@ def tests_human(priority, name, revision, ref, context, base, repo, bots_ref):
         bots_ref=bots_ref and (" [bots@%s]" % bots_ref) or "",
     )
 
+def queue_test(priority, name, revision, ref, context, base, repo, bots_ref, queue, options):
```
Comment (martinpitt) [inline bots/tests-scan]: Matter of preference, but I don't think it makes too much sense to split this into two functions.

File: bots/tests-scan [hunk 13]
```diff
@@ -173,8 +207,8 @@ def tests_invoke(priority, name, revision, ref, context, base, repo, bots_ref, o
         priority = int(priority)
     except (ValueError, TypeError):
         priority = 0
-    if priority <= 0:
-        return
+    # if priority <= 0:
+    #     return
```
Comment (martinpitt) [inline bots/tests-scan]: Why was this removed?

File: bots/tests-scan (hunk 9 above)
Comment (Gundersanne) [reply bots/tests-scan]: there's a return statement before the `with...`, as I didn't want to open the queue when called with `-vd`.

File: bots/tests-scan [hunk 14]
```diff
@@ -251,7 +277,7 @@ def prioritize(status, title, labels, priority, context):
 
         # Prefer "local" operating system
         elif os.environ.get("TEST_OS", "?") not in context:
-            priority -= random.randint(1, 2)
+            priority -= 1
```
Comment (martinpitt) [inline bots/tests-scan]: FTR, I think landing these priority computation changes now, without the queue, should be harmless enough (wrt. a possible split of this PR).

File: bots/tests-scan [hunk 15]
```diff
@@ -386,14 +417,15 @@ def scan_for_pull_tasks(api, policy, opts, repo):
...
     if opts.human_readable:
         func = lambda x: tests_human(*x)
         results.sort(reverse=True, key=lambda x: str(x))
-    else:
-        func = lambda x: tests_invoke(*x, options=opts)
-    return list(map(func, results))
+        return list(map(func, results))
+    with distributed_queue(opts.amqp, opts.queue) as queue:
+        func = lambda x: queue_test(*x, queue=queue, options=opts)
+        return list(map(func, results))
```
Comment (martinpitt) [inline bots/tests-scan]: I think this needs to get along with not specifying an amqp server/queue, and then work as it does now. This is both for an easier transition to queued, as well as developing locally and running tests-scan with `-d`.

File: bots/run-queue [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+    arguments = {
+        'x-max-priority': MAX_PRIORITY
+    }
+    channel.queue_declare(queue=opts.queue, arguments=arguments)
+    channel.basic_qos(0, 1, True)
+
+    # Get one item from the queue if present
+    msg = channel.basic_get(queue=opts.queue)
+    if msg:
+        subprocess.check_call(msg.body, shell=True)
```
[the remark this answers is not in the document]
Comment (Gundersanne) [reply bots/run-queue]: I added this in and then when I moved stuff around, I found the only thing it was using was the body so I removed the metadata again. Adding it back in would be purely for later extension, which I'm fine with also.

File: bots/tests-invoke [hunk 17]
```diff
@@ -49,6 +49,7 @@ def main():
             action='store', help='Publish results centrally to a sink')
     parser.add_argument('-v', '--verbose', action='store_true', help='Verbose output')
     parser.add_argument('context', help="The context or type of integration tests to run")
+    parser.add_argument('pull_number', help="The number of the pull request to test")
```
[the remark this answers is not in the document]
Comment (Gundersanne) [reply bots/tests-invoke]: The latter would be best I think. The collision detection based on revision is pretty important with the queues, and needs to happen during consuming, not publishing.

File: bots/make-checkout [hunk 18]
```diff
@@ -67,7 +67,10 @@ def main():
             execute("git", "remote", "remove", "test")
         execute("git", "remote", "add", "test", "https://github.com/" + opts.repo)
 
-    execute("git", "fetch", "test" if opts.repo else "origin", opts.ref)
+    # if the pr was updated while the command was in the queue, the revision
+    # won't be tied to the ref anymore, and it won't exist locally, so exit
+    # silently
+    execute("git", "fetch", "test" if opts.repo else "origin", opts.ref, permissive=True)
```
Comment (martinpitt) [inline bots/make-checkout]: My connotation of `permissive` is to ignore errors, not to abort the program with a zero exit code. But I can't think of a better name, as this is a really weird semantics for `execute()`. How about you don't use that at all, and just do `subprocess.call()` instead, and exit here on failure? (You'd need to replicate the verbose logging)

Also, this should in principle apply to the current infra as well, so could be a separate commit.

File: bots/run-queue [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+
+import argparse
+import json
+import subprocess
+import sys
+
+import amqp
+
+def main():
+    parser = argparse.ArgumentParser(description='Bot: read test commands from the queue and execute them')
```
Comment (martinpitt) [inline bots/run-queue]: It's just  processing one command, not several -- and not necessarily for tests only (at least not by design).

File: bots/run-queue [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+    connection.connect()
+    channel = connection.channel()
+    arguments = {
+        'x-max-priority': MAX_PRIORITY
+    }
+    channel.queue_declare(queue=opts.queue, arguments=arguments)
+    channel.basic_qos(0, 1, True)
+
+    # Get one item from the queue if present
+    msg = channel.basic_get(queue=opts.queue)
```
Comment (martinpitt) [inline bots/run-queue]: Just to ensure, this is non-blocking? Otherwise our simple `echo run-queue` in .tasks would not work.

File: bots/run-queue [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+    if msg:
+        body = json.loads(msg.body)
```
Comment (martinpitt) [inline bots/run-queue]: I suggest to print the command here for log purposes, at least the type and sha? But having the full JSON can't hurt for simplicity, debugging, and the ability to requeue an item manually.

File: bots/task/github.py [hunk 22]
```diff
@@ -45,6 +45,10 @@
 NOT_TESTED = "Not yet tested"
```
Comment (martinpitt) [inline bots/task/github.py]: To further explain what's going on, please add a comment here that this is a webhook-triggered test, as opposed to a user one.

File: bots/tests-invoke [hunk 23]
```diff
@@ -49,6 +49,7 @@ def main():
             action='store', help='Publish results centrally to a sink')
     parser.add_argument('-v', '--verbose', action='store_true', help='Verbose output')
     parser.add_argument('context', help="The context or type of integration tests to run")
+    parser.add_argument('pull_number', default=None, help="The number of the pull request to test")
```
Comment (martinpitt) [inline bots/tests-invoke]: I have sort of a déjà-vu here, but I think this should be an actual option `--pull-number` for clarity. Optional positional arguments are really confusing. Then you can also drop the `default=`.

To clarify, this is only being used for collision detection, right?

File: bots/tests-invoke [hunk 24]
```diff
@@ -83,11 +84,26 @@ class PullTask(object):
...
 
+        if opts.pull_number:
+            pull = api.get("pulls/{0}".format(opts.pull_number))
+            if pull and pull['head']['sha'] != self.revision:
+                return 'Newer revision available on GitHub for this pull request'
+
+        statuses = api.get('commits/' + self.github_revision + '/statuses')
+        for status in statuses:
+            if status.get('context') == self.context and status.get('state') != 'pending':
+                return 'Status of context isn\'t pending'
```
Comment (martinpitt) [inline bots/tests-invoke]: It's whimsical, but please use double quotes for strings containing a single quote.

File: bots/tests-invoke [hunk 25]
```diff
@@ -83,11 +84,26 @@ class PullTask(object):
...
         api = github.GitHub()
         if not self.github_revision:
             self.github_revision = self.revision
 
+        if opts.pull_number:
+            pull = api.get("pulls/{0}".format(opts.pull_number))
+            if pull and pull['head']['sha'] != self.revision:
+                return 'Newer revision available on GitHub for this pull request'
+
+        statuses = api.get('commits/' + self.github_revision + '/statuses')
```
Comment (martinpitt) [inline bots/tests-invoke]: Another whimsical thing, but please use `.format()` for these more complicated strings, and (possibly) `+` for the simple concatenatino in line 93 :-)

File: bots/run-queue (hunk 20 above)
Comment (Gundersanne) [reply bots/run-queue]: yup, if there's no messages in the queue this will return immediately.

File: bots/run-queue (hunk 19 above)
Comment (martinpitt) [reply bots/run-queue]: It still says "test" now, but I'll let that slip :-)

**Comment by martinpitt** [comment]
I made a check list what to look out for after deployment, and what came out of the previous test round. Please reopen after landing the revert.
